_posts/2017-07-29-ruby-oop.md

```diff
@@ -0,0 +1,381 @@
+---
+layout: post
+author: z64
+title: Beginning OOP with Ruby
+date: 2017-07-29
+---
+
+This post attempts to guide a beginner through object-oriented progamming in Ruby, showing them the thought process behind both designing and interacting with a library, and problem solving in general with Ruby.
+
+To accomplish this, we will be solving a simple geometry problem:
+
+- Describe a collection of objects that can calculate the perimeter of a polygon with an arbitrary number of sides.
+
+## Breaking it down
+
+The first thing we want to do to tackle this problem is to break apart the problem into a series of smaller, bite sized questions. "Good Ruby code" will read a lot like english - I think you'll find the syntax we end up writing will almost look like english answers to these questions. You probably already know the answers to these (which is why I picked this as an example), but read them and pay attention to the language that I use.
+
+**1) What is perimeter?**
+
+> Perimeter is the total distance around an object, in our case a 2D polygon with any number of sides.
+
+**2) What is a Polygon?**
+
+> A **polygon** is an object **made up of lines**.
+
+**3) Okay, so what is a line?**
+
+> A line is the shortest distance between **two points** in space.
+Since we're dealing with finite lines, they have a *start* and an *end*.
+
+**4) Okay, so what is a point?**
+
+> A point is a location in space. For our purposes, a location on a 2D plane, bound by **coordinates**.
+
+In the next section, we'll convert these concepts into Ruby code. Please re-read this section once we're done and see how what we said here is directly represented in the answer we come up with.
+
+## Concepts into code
+
+From the previous section, you can see that we will need three objects to describe our problem. Using the `class` keyword, we can form a basic layout:
+
+```ruby
+# A location in 2D space.
+class Point
+end
+
+# A finite segment (or distance) between two points.
+class Line
+end
+
+# A collection of lines that describe a 2D shape.
+class Polygon
+end
+```
+And from here, we can already describe how we're going to solve our problem, by creating a `#new` instance of a `Polygon`, and calling `#perimeter` on it:
+```ruby
+Polygon.new(___).perimeter
+```
+> In Ruby, and in most OOP languages, a `class` describes a "template" for an object. An object is usually a collection of data, and a set of methods or functions that allow you to interact with that data or give it logic. We first describe the objects with a `class`, and create new instances of them with `Thing.new`.
```

Comment:
`Thing.new` feels broad, maybe say like "`X.new`, where X is the name of your class"? Not sure